### ISSUE TITLE
Add SettingConveterAttribute Constructor which accepts an ISettingConverter

### ISSUE THREAD

**Issue by bretcope** [NFig/NFig#17]
Right now, the [SettingConverterAttribute](https://github.com/NFig/NFig/blob/master/NFig/SettingConverterAttribute.cs) has only one constructor. It accepts a `Type` argument, and then it creates an instance of that type and assigns it to `Converter`.

It would be more powerful if there were an additional constructor which took a pre-instantiated converter. This would allow converters of the same type, but with different options, to be instantiated and used.

The `Converter` property can also be made read-only instead of private set.

**Comment by rossipedia** [comment]
Proposed changes:

``` diff
-        public ISettingConverter Converter { get; private set; }
+        public ISettingConverter Converter { get; }

         /// <summary>
         /// Explicitly assigns a converter to a specific setting. If you want a converter to automatically apply to any setting of a particular type, pass the
@@ -20,14 +21,34 @@ namespace NFig
         /// <param name="converterType">The type must implement <see cref="ISettingConverter{TValue}"/> where TValue is the property type of the setting.</param>
         public SettingConverterAttribute(Type converterType)
         {
+            ValidateConverterType(converterType);
+            Converter = (ISettingConverter)Activator.CreateInstance(converterType);
+        }
+
+        /// <summary>
+        /// Explicitly assigns a converter to a specific setting. If you want a converter to automatically apply to any setting of a particular type, pass the
+        /// converter as part of the "additionalDefaultConverters" argument to the NFigStore you're using.
+        /// </summary>
+        /// <param name="converter">
+        /// An instance of <see cref="ISettingConverter"/>. The concrete type must implement  <see cref="ISettingConverter{TValue}"/>
+        /// where TValue is the property type of the setting.
+        /// </param>
+        protected SettingConverterAttribute([NotNull] ISettingConverter converter)
+        {
+            if (converter == null) throw new ArgumentNullException(nameof(converter));
+            ValidateConverterType(converter.GetType());
+
+            Converter = converter;
+        }
+
+        static void ValidateConverterType(Type converterType)
+        {
             // make sure type implements SettingsConverter<>
             var genericType = typeof(ISettingConverter<>);
             if (!converterType.GetInterfaces().Any(i => i.IsGenericType && i.GetGenericTypeDefinition() == genericType))
             {
                 throw new InvalidOperationException($"Cannot use type {converterType.Name} as a setting converter. It does not implement ISettingConverter<T>.");
             }
-
-            Converter = (ISettingConverter)Activator.CreateInstance(converterType);
```

See any issues?

**Comment by rossipedia** [comment]
Hmmm the diff is a little hard to follow. Here's the whole type:

``` cs
/// <summary>
    /// Used to mark individual settings as using a specific converter.
    /// </summary>
    public class SettingConverterAttribute : Attribute
    {
        /// <summary>
        /// The converter object. It will implement <see cref="ISettingConverter{TValue}"/> where TValue is the property type of the setting.
        /// </summary>
        public ISettingConverter Converter { get; }

        /// <summary>
        /// Explicitly assigns a converter to a specific setting. If you want a converter to automatically apply to any setting of a particular type, pass the
        /// converter as part of the "additionalDefaultConverters" argument to the NFigStore you're using.
        /// </summary>
        /// <param name="converterType">The type must implement <see cref="ISettingConverter{TValue}"/> where TValue is the property type of the setting.</param>
        public SettingConverterAttribute(Type converterType)
        {
            ValidateConverterType(converterType);
            Converter = (ISettingConverter)Activator.CreateInstance(converterType);
        }

        /// <summary>
        /// Explicitly assigns a converter to a specific setting. If you want a converter to automatically apply to any setting of a particular type, pass the
        /// converter as part of the "additionalDefaultConverters" argument to the NFigStore you're using.
        /// </summary>
        /// <param name="converter">
        /// An instance of <see cref="ISettingConverter"/>. The concrete type must implement  <see cref="ISettingConverter{TValue}"/> 
        /// where TValue is the property type of the setting.
        /// </param>
        protected SettingConverterAttribute([NotNull] ISettingConverter converter)
        {
            if (converter == null) throw new ArgumentNullException(nameof(converter));
            ValidateConverterType(converter.GetType());

            Converter = converter;
        }

        static void ValidateConverterType(Type converterType)
        {
            // make sure type implements SettingsConverter<>
            var genericType = typeof(ISettingConverter<>);
            if (!converterType.GetInterfaces().Any(i => i.IsGenericType && i.GetGenericTypeDefinition() == genericType))
            {
                throw new InvalidOperationException($"Cannot use type {converterType.Name} as a setting converter. It does not implement ISettingConverter<T>.");
            }
        }
    }
```

**Comment by bretcope** [comment]
That looks right to me.

Style note: there's a single-line if statement in the protected constructor.

**Comment by rossipedia** [comment]
good catch. that was auto-generated, I'll fix it up before committing
